commerce_coordinator/apps/commercetools/http_api_client.py

```diff
@@ -41,51 +41,75 @@ def _get_access_token(self) -> str:
+            Union[Dict, None]: JSON response from the API or None if all retries fail.
         """
         url = f"{self.config['apiUrl']}/{self.config['projectKey']}/{endpoint}"
         headers = {
             "Authorization": f"Bearer {self.access_token}",
             "Content-Type": "application/json",
         }
-        try:
-            response = requests.request(
-                method,
-                url,
-                headers=headers,
-                params=params,
-                json=json,
-                timeout=settings.REQUEST_READ_TIMEOUT_SECONDS,
-            )
-            response.raise_for_status()
-            return response.json()
-        except HTTPError as err:
-            if response is not None:
-                response_message = response.json().get('message', 'No message provided.')
-                logger.error(
-                    "API request for endpoint: %s failed with error: %s and message: %s",
-                    endpoint, err, response_message
+
+        def attempt(attempt_number: int) -> Union[Dict, List, None]:
+            try:
+                response = requests.request(
+                    method,
+                    url,
+                    headers=headers,
+                    params=params,
+                    json=json,
+                    timeout=settings.REQUEST_READ_TIMEOUT_SECONDS,
+                )
+
+                response.raise_for_status()
+                return response.json()
+            except requests.RequestException as err:
+                response = getattr(err, 'response', None)
+                if response is not None:
+                    try:
+                        response_message = response.json().get('message', 'No message provided.')
+                    except ValueError:
+                        response_message = response.text or 'No message provided.'
+                else:
+                    response_message = str(err) or 'No response available.'
+
+                if attempt_number >= total_retries:
+                    logger.error(
+                        "API request for endpoint: %s failed after attempt #%s with error: %s and message: %s",
+                        endpoint, attempt_number, err, response_message
+                    )
+                    return None
+
+                next_attempt = attempt_number + 1
+                next_backoff = base_backoff * next_attempt
+                logger.warning(
+                    "API request for endpoint: %s failed with error: %s and message: %s. "
+                    "Retrying attempt #%s in %s seconds...",
+                    endpoint, err, response_message, next_attempt, next_backoff
                 )
-            else:
-                logger.error("API request for endpoint: %s failed with error: %s", endpoint, err)
+                time.sleep(next_backoff)
+                return attempt(next_attempt)
```

Comment:
When the function will be terminated, I mean where is the condition that returns the flow to the main function when the retries are excited?

Reply:
its here https://github.com/edx/commerce-coordinator/pull/379/files#diff-5ac44839ca97d00c87ef9704b3c238e76ca91ae749844bb1b52f4859e0c3434cR95-R100
```
if attempt_number >= total_retries:
                    logger.error(
                        "API request for endpoint: %s failed after attempt #%s with error: %s and message: %s",
                        endpoint, attempt_number, err, response_message
                    )
                    return None
```